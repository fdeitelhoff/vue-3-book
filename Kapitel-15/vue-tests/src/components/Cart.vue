<template>
  <div>
    <h4 class="header-cart">Ihr Warenkorb</h4>
  </div>
  <div>
    <div v-for="article in allArticles" :key="article.id">
      <span style="margin-right: 10px">{{ article.name }}</span>
      <span style="margin-right: 10px">Anzahl: {{ article.quantity }}</span>
      <span style="margin-right: 10px">Gesamt: {{ article.quantity }}</span>
      <span>Kosten: {{ articlePrice(article.id) }} EUR</span>
      <div>
        <button @click="incrementQuantity(article.id)">+</button>
        <button @click="decrementQuantity(article.id)">-</button>
      </div>
    </div>
    <p>
      <span data-testId="finalPrice">Gesamtpreis: {{ finalPrice }} EUR</span>
    </p>
    <p>
      <p>
        Ihr eingegebener Gutschein ist: {{ coupon }}
      </p> 
      <label for="coupon">Gutschein:</label>
      <input v-model="coupon" id="coupon" name="coupon" />
    </p>
  </div>
</template>

<script>
export default {
  name: "Cart",
  data: function () {
    return {
      coupon: "",
    };
  },
  props: {
    articles: {
      type: Array,
      required: true,
      default: () => [],
    },
  },
  computed: {
    allArticles() {
      return this.articles;
    },
    finalPrice() {
      const finalPrice = this.articles.reduce(
        (sum, a) => sum + a.quantity * a.price,
        0
      );
      return finalPrice;
    },
  },
  methods: {
    incrementQuantity(id) {
      const article = this.articles.find((a) => a.id === id);
      article.quantity++;

      // console.log("increment quantity", id, article.quantity);
    },
    decrementQuantity(id) {
      const article = this.articles.find((a) => a.id === id);
      if (article.quantity > 1) {
        article.quantity--;
      }

      // console.log("decrement quantity", id, article.quantity);
    },
    articlePrice(id) {
      const article = this.articles.find((a) => a.id === id);
      return article.quantity * article.price;
    },
  },
};
</script>
