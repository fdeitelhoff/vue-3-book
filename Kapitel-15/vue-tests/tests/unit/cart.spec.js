import { render, fireEvent } from "@testing-library/vue";
import userEvent from "@testing-library/user-event";
import Cart from "@/components/Cart.vue";

const articles = [
  {
    id: 1,
    name: "Artikel 1",
    quantity: 2,
    price: 1.0,
  },
  {
    id: 2,
    name: "Artikel 2",
    quantity: 3,
    price: 2.0,
  },
];

describe("Cart.vue", () => {
  it("renders article names", () => {
    const { getByText } = render(Cart, {
      props: { articles },
    });

    getByText(articles[0].name);
    getByText(articles[1].name);
  });

  it("renders article quantities", () => {
    const { getByText } = render(Cart, {
      props: { articles },
    });

    getByText("Anzahl: " + articles[0].quantity);
    getByText("Anzahl: " + articles[1].quantity);
  });

  it("renders article prices", () => {
    const { getByText } = render(Cart, {
      props: { articles },
    });

    getByText("Kosten: " + articles[0].quantity * articles[0].price + " EUR");
    getByText("Kosten: " + articles[1].quantity * articles[1].price + " EUR");
  });

  it("renders initial overall price", () => {
    const { getByText } = render(Cart, {
      props: { articles },
    });

    getByText("Gesamtpreis: 8 EUR");
  });

  it("increments articles quantity", async () => {
    const { getAllByRole, getByText } = render(Cart, {
      props: { articles },
    });

    const incrementQuantityButtons = getAllByRole("button", { name: "+" });
    await fireEvent.click(incrementQuantityButtons[0]);
    await fireEvent.click(incrementQuantityButtons[1]);
    getByText("Anzahl: 3");
    getByText("Anzahl: 4");
  });

  it.skip("decrements articles quantity", async () => {
    const user = userEvent.setup();

    const { getAllByRole, getByText } = render(Cart, {
      props: { articles },
    });

    const decrementQuantityButtons = getAllByRole("button", { name: "-" });
    // console.log(decrementQuantityButtons);
    await user.click(decrementQuantityButtons[0]);
    await user.click(decrementQuantityButtons[1]);
    getByText("Anzahl: 1");
    getByText("Anzahl: 2");
  });

  it("coupon is handled properly via v-model", async () => {
    const { getByLabelText, getByText } = render(Cart, {
      props: { articles },
    });

    getByText("Ihr eingegebener Gutschein ist:");

    const couponInput = getByLabelText(/Gutschein:/i);

    await fireEvent.update(couponInput, "123456789");

    getByText("Ihr eingegebener Gutschein ist: 123456789");
  });

  it("renders the Cart.vue component correctly", () => {
    const { container } = render(Cart, {
      props: { articles },
    });

    expect(container).toMatchSnapshot();
  });
});
