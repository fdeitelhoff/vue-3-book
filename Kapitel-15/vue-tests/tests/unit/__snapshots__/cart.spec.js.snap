// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart.vue renders the Cart.vue component correctly 1`] = `
<div>
  
  <div>
    <h4
      class="header-cart"
    >
      Ihr Warenkorb
    </h4>
  </div>
  <div>
    
    <div>
      <span
        style="margin-right: 10px;"
      >
        Artikel 1
      </span>
      <span
        style="margin-right: 10px;"
      >
        Anzahl: 3
      </span>
      <span
        style="margin-right: 10px;"
      >
        Gesamt: 3
      </span>
      <span>
        Kosten: 3 EUR
      </span>
      <div>
        <button>
          +
        </button>
        <button>
          -
        </button>
      </div>
    </div>
    <div>
      <span
        style="margin-right: 10px;"
      >
        Artikel 2
      </span>
      <span
        style="margin-right: 10px;"
      >
        Anzahl: 4
      </span>
      <span
        style="margin-right: 10px;"
      >
        Gesamt: 4
      </span>
      <span>
        Kosten: 8 EUR
      </span>
      <div>
        <button>
          +
        </button>
        <button>
          -
        </button>
      </div>
    </div>
    
    <p>
      <span
        data-testid="finalPrice"
      >
        Gesamtpreis: 11 EUR
      </span>
    </p>
    <p>
      <p>
         Ihr eingegebener Gutschein ist: 
      </p>
      <label
        for="coupon"
      >
        Gutschein:
      </label>
      <input
        id="coupon"
        name="coupon"
      />
    </p>
  </div>
  
</div>
`;
